test/jdk/java/net/httpclient/SpecialHeadersTest.java

```diff
@@ -167,48 +320,54 @@ void test(String uriString,
-                out.println("Got X-" + key + ": " + resp.headers().allValues("X-"+key));
-                if (value != null) {
-                    assertEquals(receivedHeaderString, value);
-                    assertEquals(resp.headers().allValues("X-"+key), List.of(value));
+            try {
+                if (!sameClient || client == null)
+                    client = newHttpClient("test", sameClient);
+
+                HttpRequest.Builder requestBuilder = HttpRequest.newBuilder(uri);
+                if (!useDefault) {
+                    requestBuilder.header(name, value);
+                }
+                HttpRequest request = requestBuilder.build();
+                HttpResponse<String> resp = client.send(request, BodyHandlers.ofString());
+
+                out.println("Got response: " + resp);
+                out.println("Got body: " + resp.body());
+                assertEquals(resp.statusCode(), 200,
+                        "Expected 200, got:" + resp.statusCode());
+
+                boolean isInitialRequest = i == 0;
+                boolean isSecure = uri.getScheme().equalsIgnoreCase("https");
+                boolean isHTTP2 = resp.version() == HTTP_2;
+                boolean isNotH2CUpgrade = isSecure || (sameClient == true && !isInitialRequest);
+                boolean isDefaultHostHeader = name.equalsIgnoreCase("host") && useDefault;
+
+                // By default, HTTP/2 sets the `:authority:` pseudo-header, instead
+                // of the `Host` header. Therefore, there should be no "X-Host"
+                // header in the response, except the response to the h2c Upgrade
+                // request which will have been sent through HTTP/1.1.
+
+                if (isDefaultHostHeader && isHTTP2 && isNotH2CUpgrade) {
+                    assertTrue(resp.headers().firstValue("X-" + key).isEmpty());
+                    assertTrue(resp.headers().allValues("X-" + key).isEmpty());
+                    out.println("No X-" + key + " header received, as expected");
                 } else {
-                    assertEquals(resp.headers().allValues("X-"+key).size(), 0);
+                    String receivedHeaderString = value == null ? null
+                            : resp.headers().firstValue("X-" + key).orElse(null);
+                    out.println("Got X-" + key + ": " + resp.headers().allValues("X-" + key));
+                    if (value != null) {
+                        assertEquals(receivedHeaderString, value);
+                        assertEquals(resp.headers().allValues("X-" + key), List.of(value));
+                    } else {
+                        assertEquals(resp.headers().allValues("X-" + key).size(), 0);
+                    }
+                }
+            } finally {
+                if (!sameClient) {
+                    client = null;
+                    System.gc();
+                    var error = TRACKER.check(500);
+                    if (error != null) throw error;
```

Comment:
Given slowness on CI systems, would it be better to use the jtreg timeout factor here to adjust the `500ms` timeout? Or better still, perhaps just log a warning instead of throwing an error? After all this test's main goal is to verify the request headers and in that context it should be OK if some HttpClient(s) aren't gc collected by the time the test completes?

Reply:
I haven't observed any issue yet with using a `500ms`. Here or in other tests. And I do want to throw an error if a client doesn't cleanly exits. So if we get a failure here - we should look at what prevented the client from exiting - and if it looks like it might be that the GC didn't get enough cycle then we would raise the timeout.